src/drivers/fluid_pipewire.c

```diff
@@ -0,0 +1,277 @@
+
+    pw_stream_queue_buffer(drv->pw_stream, pwb);
+}
+
+/* Rendering routine with support for user-defined audio manipulation */
+static void fluid_pipewire_event_process2(void *data)
+{
+    fluid_pipewire_audio_driver_t *drv = data;
+    struct pw_buffer *pwb;
+    struct spa_buffer *buf;
+    float *dest;
+    float *channels[NUM_CHANNELS] = { drv->lbuf, drv->rbuf };
+    int stride, i;
+
+    pwb = pw_stream_dequeue_buffer(drv->pw_stream);
+    if (!pwb)
+    {
+        FLUID_LOG(FLUID_WARN, "No buffers!");
+        return;
+    }
+
+    buf = pwb->buffer;
+    dest = buf->datas[0].data;
+    if (!dest)
+    {
+        return;
+    }
+
+    stride = sizeof(float) * NUM_CHANNELS;
+
+    FLUID_MEMSET(drv->lbuf, 0, drv->buffer_period * sizeof(float));
+    FLUID_MEMSET(drv->rbuf, 0, drv->buffer_period * sizeof(float));
+
+    (*drv->user_callback)(drv->data, drv->buffer_period, 0, NULL, NUM_CHANNELS, channels);
+
+    /* Interleave the floating point data */
+    for (i = 0; i < drv->buffer_period; i++)
+    {
+        dest[i * 2] = drv->lbuf[i];
+        dest[i * 2 + 1] = drv->rbuf[i];
+    }
+
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->stride = stride;
+    buf->datas[0].chunk->size = drv->buffer_period * stride;
+
+    pw_stream_queue_buffer(drv->pw_stream, pwb);
+}
+
+fluid_audio_driver_t *new_fluid_pipewire_audio_driver(fluid_settings_t *settings, fluid_synth_t *synth)
+{
+    return new_fluid_pipewire_audio_driver2(settings, NULL, synth);
+}
+
+static struct pw_stream_events events = {
+    PW_VERSION_STREAM_EVENTS
+};
+
+fluid_audio_driver_t *
+new_fluid_pipewire_audio_driver2(fluid_settings_t *settings, fluid_audio_func_t func, void *data)
+{
+    fluid_pipewire_audio_driver_t *drv;
+    int period_size;
+    int buffer_length;
+    int res;
+    int pw_flags;
+    int realtime_prio = 0;
+    double sample_rate;
+    char *media_role = NULL;
+    char *media_type = NULL;
+    char *media_category = NULL;
+    float *buffer = NULL;
+    const struct spa_pod *params[1];
+
+    drv = FLUID_NEW(fluid_pipewire_audio_driver_t);
+    if (!drv)
+    {
+        FLUID_LOG(FLUID_ERR, "Out of memory");
+        return NULL;
+    }
+
+    FLUID_MEMSET(drv, 0, sizeof(fluid_pipewire_audio_driver_t));
+
+    fluid_settings_getint(settings, "audio.period-size", &period_size);
+    fluid_settings_getint(settings, "audio.realtime-prio", &realtime_prio);
+    fluid_settings_getnum(settings, "synth.sample-rate", &sample_rate);
+    fluid_settings_dupstr(settings, "audio.pipewire.media-role", &media_role);
+    fluid_settings_dupstr(settings, "audio.pipewire.media-type", &media_type);
+    fluid_settings_dupstr(settings, "audio.pipewire.media-category", &media_category);
+
+    drv->data = data;
+    drv->user_callback = func;
+    drv->buffer_period = period_size;
+
+    events.process = func ? fluid_pipewire_event_process2 : fluid_pipewire_event_process;
+
+    pw_init(NULL, NULL);
```

Comment:
This is dangerous to do in here. When creating the driver multiple times, it would init multiple times. And every time it would change some static variables in global memory. I suggest to completely eleminate this call and leave it to downstream apps to properly init() and deinit(), like we decided to do for SDL2 driver.

Reply:
How will we access the init and deinit functions downstream?

Reply:
You'll have to include the header.

Reply:
Done, will need to find a way to detect that pw hasn't been init.

Reply:
oh